src/vsock/src/thread_backend.rs

```diff
@@ -0,0 +1,242 @@
+            Some(conn) => conn,
+            None => {
+                // assume that the connection does not exist
+                return Ok(());
+            }
+        };
+
+        if conn.rx_queue.peek() == Some(RxOps::Reset) {
+            // Handle RST events here
+            let conn = self.conn_map.remove(&key).unwrap();
+            self.listener_map.remove(&conn.stream.as_raw_fd());
+            self.stream_map.remove(&conn.stream.as_raw_fd());
+            self.local_port_set.remove(&conn.local_port);
+            VhostUserVsockThread::epoll_unregister(conn.epoll_fd, conn.stream.as_raw_fd())
+                .unwrap_or_else(|err| {
+                    warn!(
+                        "Could not remove epoll listener for fd {:?}: {:?}",
+                        conn.stream.as_raw_fd(),
+                        err
+                    )
+                });
+
+            // Initialize the packet header to contain a VSOCK_OP_RST operation
+            pkt.set_op(VSOCK_OP_RST)
+                .set_src_cid(VSOCK_HOST_CID)
+                .set_dst_cid(conn.guest_cid)
+                .set_src_port(conn.local_port)
+                .set_dst_port(conn.peer_port)
+                .set_len(0)
+                .set_type(VSOCK_TYPE_STREAM)
+                .set_flags(0)
+                .set_buf_alloc(0)
+                .set_fwd_cnt(0);
+
+            return Ok(());
+        }
+
+        // Handle other packet types per connection
+        conn.recv_pkt(pkt)?;
+
+        Ok(())
+    }
+
+    /// Deliver a guest generated packet to its destination in the backend.
+    ///
+    /// Absorbs unexpected packets, handles rest to respective connection
+    /// object.
+    ///
+    /// Returns:
+    /// - always `Ok(())` if packet has been consumed correctly
+    pub fn send_pkt(&mut self, pkt: &VsockPacket) -> Result<()> {
+        let key = ConnMapKey::new(pkt.dst_port(), pkt.src_port());
+
+        // TODO: Rst if packet has unsupported type
+        if pkt.pkt_type() != VSOCK_TYPE_STREAM {
+            info!("vsock: dropping packet of unknown type");
+            return Ok(());
+        }
+
+        // TODO: Handle packets to other CIDs as well
+        if pkt.dst_cid() != VSOCK_HOST_CID {
+            info!(
+                "vsock: dropping packet for cid other than host: {:?}",
+                pkt.hdr()
+            );
+
+            return Ok(());
+        }
+
+        // TODO: Handle cases where connection does not exist and packet op
+        // is not VSOCK_OP_REQUEST
+        if !self.conn_map.contains_key(&key) {
+            // The packet contains a new connection request
+            if pkt.op() == VSOCK_OP_REQUEST {
+                self.handle_new_guest_conn(&pkt);
+            } else {
+                // TODO: send back RST
+            }
+            return Ok(());
+        }
+
+        if pkt.op() == VSOCK_OP_RST {
+            // Handle an RST packet from the guest here
```

Comment:
is it possible to invert the logic above and do a match on pkt.op() and cleanly handle the various ops. You could test for the key and handle the result inside the VSOCK_OP_REQUEST => match.

Reply:
@stsquad I am not sure about that. Since the VsockThreadBackend contains the state information of the current connections and since we need to add/delete connections when we get a VSOCK_OP_REQUEST or VSOCK_OP_RST, respectively, I think it makes sense to do it here, let me know what you think. However, I changed the per connection (VsockConnection) send_pkt to be implemented using a match.